module/validation/receipt_validator.go

```diff
@@ -148,29 +149,152 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
 // Expected errors during normal operations:
 // * engine.InvalidInputError
 // * validation.UnverifiableError
-func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
+func (v *receiptValidator) Validate(receipt *flow.ExecutionReceipt) error {
+	// TODO: this can be optimized by checking if result was already stored and validated.
+	// This needs to be addressed later since many tests depend on this behavior.
+	prevResult, err := v.fetchResult(receipt.ExecutionResult.PreviousResultID)
+	if err != nil {
+		return err
+	}
+
+	// first validate result to avoid signature check in in `validateReceipt` in case result is invalid.
+	err = v.validateResult(&receipt.ExecutionResult, prevResult)
+	if err != nil {
+		return fmt.Errorf("could not validate single result %v at index: %w", receipt.ExecutionResult.ID(), err)
+	}
+
+	err = v.validateReceipt(receipt.Meta(), receipt.ExecutionResult.BlockID)
+	if err != nil {
+		// It's very important that we fail the whole validation if one of the receipts is invalid.
+		// It allows us to make assumptions as stated in previous comment.
+		return fmt.Errorf("could not validate single receipt %v: %w", receipt.ID(), err)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
 	// lookup cache to avoid linear search when checking for previous result that is
 	// part of payload
-	payloadExecutionResults := make(map[flow.Identifier]*flow.ExecutionResult)
-	for _, receipt := range receipts {
-		payloadExecutionResults[receipt.ExecutionResult.ID()] = &receipt.ExecutionResult
-	}
+	payloadExecutionResults := candidate.Payload.ResultsById()
 	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
 	// local storage. This is needed to handle a case when same block payload contains receipts that
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
```

Comment:
⚠️ 
I think there are two uncovered edge cases here in the logic
1. We would except a receipt with a missing result: 
    - As long as the result is known `fetchResult` is happy. For example, the result could be included in another fork (or otherwise known to the consensus node).  As far as I understand, there is no check that the result has been included in the fork (?)
2. Consider an execution fork that conflicts with the latest finalized block on this fork. I believe the current logic would happily accept results and receipts for the conflicting fork. 

I have updated the test to demonstrate this behaviour: 
https://github.com/onflow/flow-go/blob/48610ee637e030da031f5b1484c0d1f60413f280/module/validation/receipt_validator_test.go#L472
(caution: the updated test is on a different branch).

Reply:
Yes, that is true, but that is the case with current implementation as well. I was going to address this in separate PR, since it's tightly related with https://github.com/dapperlabs/flow-go/issues/5253. Of course I can do this here as well.